File: BeerViewer/Models/Materials.cs
using BeerViewer.Network;
using BeerViewer.Models.Raw;
using BeerViewer.Models.kcsapi;

namespace BeerViewer.Models
{
	public class Materials : Notifier
	{
		#region Fuel Property
		private int _Fuel;
		public int Fuel
		{
			get { return this._Fuel; }
			private set
			{
				if (this._Fuel != value)
				{
					this._Fuel = value;
					this.RaisePropertyChanged();
				}
			}
		}
		#endregion

		#region Ammo Property
		private int _Ammo;
		public int Ammo
		{
			get { return this._Ammo; }
			private set
			{
				if (this._Ammo != value)
				{
					this._Ammo = value;
					this.RaisePropertyChanged();
				}
			}
		}
		#endregion

		#region Steel Property
		private int _Steel;
		public int Steel
		{
			get { return this._Steel; }
			private set
			{
				if (this._Steel != value)
				{
					this._Steel = value;
					this.RaisePropertyChanged();
				}
			}
		}
		#endregion

		#region Bauxite Property
		private int _Bauxite;
		public int Bauxite
		{
			get { return this._Bauxite; }
			private set
			{
				if (this._Bauxite != value)
				{
					this._Bauxite = value;
					this.RaisePropertyChanged();
				}
			}
		}
		#endregion

		#region DevelopmentMaterial Property
		private int _DevelopmentMaterial;
		public int DevelopmentMaterial
		{
			get { return this._DevelopmentMaterial; }
			private set
			{
				if (this._DevelopmentMaterial != value)
				{
					this._DevelopmentMaterial = value;
					this.RaisePropertyChanged();
				}
			}
		}
		#endregion

		#region RepairBucket Property
		private int _RepairBucket;
		public int RepairBucket
		{
			get { return this._RepairBucket; }
			private set
			{
				if (this._RepairBucket != value)
				{
					this._RepairBucket = value;
					this.RaisePropertyChanged();
				}
			}
		}
		#endregion

		#region InstantConstruction Property
		private int _InstantConstruction;
		public int InstantConstruction
		{
			get { return this._InstantConstruction; }
			private set
			{
				if (this._InstantConstruction != value)
				{
					this._InstantConstruction = value;
					this.RaisePropertyChanged();
				}
			}
		}
		#endregion

		#region ImprovementMaterial Property
		private int _ImprovementMaterial;
		public int ImprovementMaterial
		{
			get { return this._ImprovementMaterial; }
			set
			{
				if (this._ImprovementMaterial != value)
				{
					this._ImprovementMaterial = value;
					this.RaisePropertyChanged();
				}
			}
		}
		#endregion


		internal Materials()
		{
			var proxy = Proxy.Instance;

			proxy.Register<kcsapi_material[]>(Proxy.api_get_member_material, x => this.Update(x.Data));
			proxy.Register<kcsapi_charge>(Proxy.api_req_hokyu_charge, x => this.Update(x.Data.api_material));
			proxy.Register<kcsapi_destroyship>(Proxy.api_req_kousyou_destroyship, x => this.Update(x.Data.api_material));
			proxy.Register<kcsapi_destroyitem2>(Proxy.api_req_kousyou_destroyitem2, x => this.Update(x.Data.api_material));
		}

		internal void Update(kcsapi_material[] source)
		{
			if (source != null && source.Length >= 8)
			{
				this.Fuel = source[0].api_value;
				this.Ammo = source[1].api_value;
				this.Steel = source[2].api_value;
				this.Bauxite = source[3].api_value;
				this.DevelopmentMaterial = source[6].api_value;
				this.RepairBucket = source[5].api_value;
				this.InstantConstruction = source[4].api_value;
				this.ImprovementMaterial = source[7].api_value;
			}
		}
		private void Update(int[] source)
		{
			if (source != null && source.Length == 4)
			{
				this.Fuel = source[0];
				this.Ammo = source[1];
				this.Steel = source[2];
				this.Bauxite = source[3];
			}
		}
	}
}

File: WindowFrame/src/KanColle/Classes/Materials.ts
import { Component } from "vue-property-decorator";
import Proxy from "@/Proxy/Proxy";

import { ParseKcsApi, } from "@KC/Functions";
import { kcsapi_port } from "@KC/Raw/kcsapi_port";
import { kcsapi_material } from "@KC/Raw/kcsapi_material";
import { kcsapi_charge } from "@KC/Raw/kcsapi_charge";
import { kcsapi_destroyship } from "@KC/Raw/kcsapi_ship";
import { kcsapi_destroyitem2 } from "../Raw/kcsapi_item";
import { kcsapi_req_nyukyo_start } from "../Raw/kcsapi_repair";
import { kcsapi_airbase_corps_supply, kcsapi_airbase_corps_set_plane, kcsapi_req_air_corps_set_plane } from "../Raw/kcsapi_airbase_corps";
import KanColleStoreClient, { StoreInterface } from "../Store/KanColleStoreClient";

/**
 * Intercept and Manage `Materials` data of KanColle
 */
@Component
export default class Materials extends KanColleStoreClient {
	constructor() {
		super();

		Proxy.Instance.Register("/kcsapi/api_port/port", (req, resp) => {
			ParseKcsApi<kcsapi_port, unknown>(resp, req.body, (x, y) => {
				this.Update(x.api_material);
			});
		});
		Proxy.Instance.Register("/kcsapi/api_get_member/material", (req, resp) => {
			ParseKcsApi<kcsapi_material[]>(resp, (x) => this.Update(x));
		});
		Proxy.Instance.Register("/kcsapi/api_req_hokyu/charge", (req, resp) => {
			ParseKcsApi<kcsapi_charge>(resp, (x) => this.Update(x.api_material));
		});
		Proxy.Instance.Register("/kcsapi/api_req_kousyou/destroyship", (req, resp) => {
			ParseKcsApi<kcsapi_destroyship>(resp, (x) => this.Update(x.api_material));
		});
		Proxy.Instance.Register("/kcsapi/api_req_hokyu/charge", (req, resp) => {
			ParseKcsApi<kcsapi_charge>(resp, (x) => this.Update(x.api_material));
		});
		Proxy.Instance.Register("/kcsapi/api_req_kousyou/destroyitem2", (req, resp) => {
			ParseKcsApi<kcsapi_destroyitem2>(resp, (x) => this.Update(x.api_get_material));
		});

		// Bucket using
		Proxy.Instance.Register("/kcsapi/api_req_nyukyo/start", (req, resp) => {
			ParseKcsApi<{}, kcsapi_req_nyukyo_start>(resp, (req as any).body as Buffer, (_, x) => (x.api_highspeed === 1) && this.DecreaseBucket());
		});
		Proxy.Instance.Register("/kcsapi/api_req_nyukyo/speedchange", (req, resp) => {
			ParseKcsApi(resp, () => this.DecreaseBucket());
		});

		// Supply Airbase
		Proxy.Instance.Register("/kcsapi/api_req_air_corps/supply", (req, resp) => {
			ParseKcsApi<kcsapi_airbase_corps_supply>(resp, (x) => this.Update({
				Fuel: x.api_after_fuel,
				Bauxite: x.api_after_bauxite
			}));
		});

		// Set aircraft to AirBase
		Proxy.Instance.Register("/kcsapi/api_req_air_corps/supply", (req, resp) => {
			ParseKcsApi<kcsapi_airbase_corps_set_plane, kcsapi_req_air_corps_set_plane>(resp, (req as any).body as Buffer, (x, y) => {
				if (y.api_item_id === -1) return;
				if (x.api_plane_info.length >= 2) return;
				this.Update({ Bauxite: x.api_after_bauxite });
			});
		});
	}

	private DecreaseBucket() {
		this.StoreUpdateMaterials({
			Bucket: this.StoreMaterials.Bucket - 1,
		});
	}

	private Update(source: number[] | kcsapi_material[] | StoreInterface.MaterialsPayload): void {
		if (Array.isArray(source)) {
			if (!source || source.length === 0) return;
			const type = typeof (source[0] as kcsapi_material).api_value;

			if (source.length >= 4 && type === "undefined") {
				const casted = source as [number, number, number, number];
				this.StoreUpdateMaterials({
					Fuel: casted[0],
					Ammo: casted[1],
					Steel: casted[2],
					Bauxite: casted[3],
				});
			} else if (source.length >= 8 && type !== "undefined") {
				const casted = source as kcsapi_material[];
				this.StoreUpdateMaterials({
					Fuel: casted[0].api_value,
					Ammo: casted[1].api_value,
					Steel: casted[2].api_value,
					Bauxite: casted[3].api_value,
					Bucket: casted[4].api_value,
					Construction: casted[5].api_value,
					Development: casted[6].api_value,
					Improvement: casted[7].api_value,
				});
			}
		} else
			this.StoreUpdateMaterials(source);
	}
}
